packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,893 @@
+    // restoration buckets in the [Widget] tree because of the way widgets are
+    // disposed during the build phase: Flutter delays the disposal of widgets
+    // until the very end of the build phase. When during a build widget A is
+    // getting replaced by widget B, B will build before A is disposed. This
+    // causes a problem if both widgets want to claim a child restoration bucket
+    // with ID x. In theory, this should be allowed because since A is getting
+    // replaced by B, A will give up the bucket in its dispose method so that B
+    // can claim a bucket with the same ID when it builds. However, in reality B
+    // will claim the bucket before A has released it, which means for a very
+    // short time (until A is actually getting disposed) there will be two child
+    // buckets under the same restoration ID, which is illegal. To get around
+    // this problem, restoration buckets do actually allow claiming child
+    // buckets for restoration IDs that have already been claimed in the hopes
+    // that the original owner will give up the bucket before the restoration
+    // data is finalized and sent to the engine at the end of the frame. They
+    // check whether this hope became reality in a finalizer callback.
+
+    assert(_rootBucket != null);
+    assert(!_debugDoingUpdate, 'Calling scheduleUpdate from a finalizer is not allowed.');
+    if (finalizer != null) {
+      _finalizers.add(finalizer);
+    }
+    if (!_postFrameScheduled) {
+      _postFrameScheduled = true;
+      SchedulerBinding.instance.addPostFrameCallback((Duration _) => _doProcessing());
+    }
+  }
+
+  // TODO(goderbauer): Add API to request out-of-band serialization when no frame is scheduled.
+
+  void _doProcessing() {
+    assert(() {
+      _debugDoingUpdate = true;
+      return true;
+    }());
+    _postFrameScheduled = false;
+
+    for (final VoidCallback finalizer in _finalizers) {
+      finalizer();
+    }
+    _finalizers.clear();
+    sendToEngine(_rootBucket._rawData);
+
+    assert(() {
+      _debugDoingUpdate = false;
+      return true;
+    }());
+  }
+}
+
+/// Identifies a piece of data within or a child of a [RestorationBucket].
+///
+/// The [value] of a [RestorationId]s must be unique within a given
+/// [RestorationBucket]. [RestorationId] objects with the same [value] are
+/// considered equal.
+@immutable
+class RestorationId {
+  /// Creates a [RestorationId].
+  ///
+  /// The provided `value` must not be null.
+  const RestorationId(this.value) : assert(value != null);
+
+  /// The value of the [RestorationId].
+  ///
+  /// Two [RestorationId] objects with the same [value] are considered equal.
+  final String value;
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType)
+      return false;
+    return other is RestorationId && other.value == value;
+  }
+
+  @override
+  int get hashCode => hashValues(runtimeType, value);
```

Comment:
Hashing the runtime type seems strange - why are we doing that? Can you document it here?

Reply:
Removed that, I think this was left-over from copying this from elsewhere.